obelisk/cpp/zoo/hardware/sensing/zed/zed2_sensors_interface.h

```diff
@@ -0,0 +1,420 @@
+                return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::FAILURE;
+            }
+
+            RCLCPP_INFO(this->get_logger(), "Camera with serial number %s initialized!",
+                        cam_param_dict.at("serial_number").c_str());
+
+            cam_images_[cam_index] = sl::Mat();
+
+            // Set runtime parameters
+            sl::RuntimeParameters rtp;
+            rtp.enable_depth          = depth_;
+            cam_rt_params_[cam_index] = rtp;
+        }
+
+        // Create timers for each camera
+        cam_cbg_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
+        for (const auto& cam_param : cam_param_dicts_) {
+            int cam_index = cam_param.first;
+
+            auto timer_callback = [this, cam_index]() { this->camera_callback(cam_index); };
+
+            double timer_period = 1.0 / static_cast<double>(fps_);
+            auto timer = this->create_wall_timer(std::chrono::duration<double>(timer_period), timer_callback, cam_cbg_);
+            std::string timer_key = "timer_" + std::to_string(cam_index);
+            timers_[timer_key]    = timer;
+        }
+
+        // Create a timer for the publisher
+        auto pub_timer_callback = [this]() { this->publish_images(); };
+        if (pub_period_ == 0) {
+            pub_period_ = 1.0 / static_cast<double>(fps_);
+        }
+        auto pub_timer =
+            this->create_wall_timer(std::chrono::duration<double>(pub_period_), pub_timer_callback, cam_cbg_);
+        std::string pub_timer_key = "timer_pub";
+        timers_[pub_timer_key]    = pub_timer;
+
+        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+    }
+
+    /**
+     * @brief Callback for shutting down the node.
+     *
+     * @param prev_state Previous state of the lifecycle node.
+     * @return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+     * Callback return indicating success or failure.
+     */
+    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+    on_shutdown(const rclcpp_lifecycle::State& prev_state) {
+        obelisk::ObeliskSensor::on_shutdown(prev_state);
+        for (auto& cam : cams_) {
+            cam.second.close();
+        }
+        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+    }
+
+  private:
+    // Key for the image publisher
+    std::string const pub_img_key_ = "pub_img";
+    double pub_period_;
+
+    // Callback group shared by all camera callbacks
+    rclcpp::CallbackGroup::SharedPtr cam_cbg_;
+
+    // Camera parameter maps
+    std::map<int, std::map<std::string, std::string>> cam_param_dicts_;
+    std::map<int, sl::Camera> cams_;
+    std::map<int, sl::Mat> cam_images_;
+    std::map<int, sl::RuntimeParameters> cam_rt_params_;
+    std::map<int, Eigen::Tensor<uint8_t, 3>> frames_;
```

Comment:
did you want to change these to vectors, or is there a reason they are maps?

Reply:
so here's the logic: I originally had these be strings (camera names), but then I realized that there was no way to guarantee their order if just indexed by the string name. Thus, I changed the names to integers. In loops over these maps, the values are drawn by order of their keys, which means that technically, the names don't have to be things like 0, 1, ... It can be something like 51, 73, ..., which is stupid, but doesn't impose any restrictions on the end user filling out the parameter yaml file.

If you feel strongly about this, we can change all these maps to vectors and I can change the yaml formatting too, though... I might be able to make it a list so there aren't even keys associated with the camera settings.

Reply:
as per f2f: changed all maps to std::vectors and refactored code to work with vectors. to facilitate this change, the camera parameter yaml file is now a list instead of a dict with int keys.